<!DOCTYPE html><html><div class="item-title">
        Item 153
      </div> <div class="item-details"><div><b>git_comments:</b> <ol></ol></div><div><b>git_commits:</b> <ol><li><div><div><b>summary:</b> HDFS-2451. TestNodeCount.testNodeCount failes with NPE. Contributed by Konstantin Boudnik.
                </div><div><b>message:</b> HDFS-2451. TestNodeCount.testNodeCount failes with NPE. Contributed by Konstantin Boudnik.


git-svn-id: https://svn.apache.org/repos/asf/hadoop/common/branches/branch-0.22@1186556 13f79535-47bb-0310-9956-ffa450edef68

                </div></div></li></ol></div><div><b>github_issues:</b> <ol></ol></div><div><b>github_issues_comments:</b> <ol></ol></div><div><b>github_pulls:</b> <ol></ol></div><div><b>github_pulls_comments:</b> <ol></ol></div><div><b>github_pulls_reviews:</b> <ol></ol></div><div><b>jira_issues:</b> <ol><li><div><div><b>summary:</b> TestNodeCount.testNodeCount failes with NPE
                </div><div><b>description:</b> in the [commit build #97|http://is.gd/UPuXg2] TestNodeCount.testNodeCount failed with NPE
{noformat}
java.lang.NullPointerException
	at org.apache.hadoop.hdfs.server.namenode.BlockManager.countNodes(BlockManager.java:1436)
	at org.apache.hadoop.hdfs.server.namenode.TestNodeCount.__CLR2_4_39bdgm6whp(TestNodeCount.java:119)
	at org.apache.hadoop.hdfs.server.namenode.TestNodeCount.testNodeCount(TestNodeCount.java:40)

{noformat}
                </div></div></li></ol></div><div><b>jira_issues_comments:</b> <ol><li><div>
                Seems like the call to {{blockManager.countNodes()}} is happening under namesystem's write locking elsewhere. Otherwise a race condition is likely.
              </div></li><li><div><div><b>body:</b> Hi Cos,
 I too think that was the cause. There is one more place in the test, where it is not taking the lock on fsnamesystem.
{code}
 // The block should be replicated
      do {
        num = namesystem.blockManager.countNodes(block);
      } while (num.liveReplicas() != REPLICATION_FACTOR);
      
{code}
Do you have any reason for not putting lock here?

Thanks
Uma
                </div><div><b>label:</b> code-design
                </div></div></li><li><div>
                Because I've missed it ;( Thanks for catching.
              </div></li><li><div>
                Looks like a race condition going on in the test itself. BlockManager.countNodes() should be called under a lock (readLock() TestNodeCount). If not under lock somebody can remove a replica between nodeIter.hasNext() and nodeIter.next() in countNodes(), which will return NULL and cause NPE.
              </div></li><li><div>
                Sorry for the late comment. I wrote it yesterday, and pushed the Add button only today... The patch looks good except that ".gitignore" changes are unrelated.
              </div></li><li><div>
                Yeah, .gitignore is relevant yet useful ;)
Here's new one.
I will run commit the patch soon
              </div></li><li><div>
                I have committed it to 0.22 branch.

An interesting observation: this problem has been fixed in the trunk a while ago as I have found out just yet checking if the patch needs to be applied for the trunk as well. Apparently, backporting wasn't a priority for trunk development ;(
              </div></li><li><div>
                As I suspected: it has been fixed long time ago in the trunk ;(
              </div></li><li><div>
                And one more....
              </div></li><li><div>
                Integrated in Hadoop-Hdfs-22-branch #100 (See [https://builds.apache.org/job/Hadoop-Hdfs-22-branch/100/])
    HDFS-2451. TestNodeCount.testNodeCount failes with NPE. Contributed by Konstantin Boudnik.

cos : http://svn.apache.org/viewcvs.cgi/?root=Apache-SVN&amp;view=rev&amp;rev=1186556
Files : 
* /hadoop/common/branches/branch-0.22/hdfs/CHANGES.txt
* /hadoop/common/branches/branch-0.22/hdfs/src/test/hdfs/org/apache/hadoop/hdfs/server/namenode/TestNodeCount.java

              </div></li></ol></div></div></html>